cpp/src/parquet/arrow/reader.cc

```diff
@@ -199,10 +201,11 @@ class FileReaderImpl : public FileReader {
     return ReadRowGroups(Iota(reader_->metadata()->num_row_groups()), indices, out);
   }
 
-  Status GetFieldReader(int i,
-                        const std::shared_ptr<std::unordered_set<int>>& included_leaves,
-                        const std::vector<int>& row_groups,
-                        std::unique_ptr<ColumnReaderImpl>* out) {
+  Status GetFieldReader(
```

Comment:
This function signature now looks strange to me since it contains conflicting parameters `const std::vector<int>& row_groups` and `const std::shared_ptr<std::vector<std::unique_ptr<RowRanges>>>& row_ranges_per_rg`.

What about splitting file-based row ranges lazily? Then we can re-define this as 
```
  // RowGroups can be either std::vector<int> or RowRanges
  template <typename RowGroups>
  Status GetFieldReader(int i,
                        const std::shared_ptr<std::unordered_set<int>>& included_leaves,
                        const RowGroups& row_groups,
                        std::unique_ptr<ColumnReaderImpl>* out);
```

Or if you still want to split file-based row ranges eagerly, we can do this:
```
  // RowGroups can be either std::vector<int> or std::map<int, RowRanges>
  template <typename RowGroups>
  Status GetFieldReader(int i,
                        const std::shared_ptr<std::unordered_set<int>>& included_leaves,
                        const RowGroups& row_groups,
                        std::unique_ptr<ColumnReaderImpl>* out);
```

Reply:
Then we can limit the scope of refactering work to overload SomeRowGroupsFactory below at line 221:
```
ctx->iterator_factory = SomeRowGroupsFactory(row_groups);
```